File: desktop/Polygon.cpp
/* Polygon.cpp
Copyright 2020 Michael Zahniser
*/

#include "Polygon.h"

#include "Edge.h"

#include <algorithm>
#include <cstdint>
#include <utility>

using namespace std;

namespace {
	// Get the maximum possible order value for a given ring.
	int MaxOrder(const Ring &ring)
	{
		Edge edge(ring);
		while(edge)
			++edge;
		return edge.Order();
	}
	
	// An Intersection is a point where P and Q intersect. Depending on what
	// algorithm step we're in, the data members are used slightly differently.
	class Intersection {
	public:
		Intersection() = default;
		Intersection(Point point, int order, int index = 0, bool entering = false)
			: point(point), order(order), index(index), entering(entering) {}
		
		// Comparison operator, which will sort intersections by order.
		bool operator<(const Intersection &other) const { return order < other.order; }
		
		// Coordinates of the intersection.
		Point point;
		// A distance value to be used to sort intersections on the same edge.
		int order;
		// The index of the associated vertex: the source vertex, if this is in
		// the first algorithm stage, or the link, in the second stage.
		int index;
		// Whether P is entering Q (as opposed to exiting it) at this point.
		bool entering;
	};
	
	// Sort a list of intersections.
	void Sort(vector<Intersection> &intersections)
	{
		sort(intersections.begin(), intersections.end());
	}
	
	// Any time two intersections occur at the same point, that means a vertex
	// touched but the edge "turned back" and did not pass through the polygon.
	// Those intersections should be erased.
	void RemoveDoubles(vector<Intersection> &intersections)
	{
		size_t j = 0;
		for(size_t i = 0; i < intersections.size(); )
		{
			// As we go, we'll shift the intersections we preserve into the
			// start of the vector.
			intersections[j] = intersections[i];
			// Find all intersections in the list that have the same order (i.e.
			// the same location) as this one. If there are multiples, we will
			// remove all of them if half are entering and half are not, or
			// all but one of them if all or none of them are entering.
			// This gets rid of corners that touch an edge without crossing it.
			int order = intersections[i].order;
			int entering = intersections[i].entering;
			int total = 1;
			for(++i; i < intersections.size() && intersections[i].order == order; ++i)
			{
				entering += intersections[i].entering;
				++total;
			}
			
			// Preserve this intersection if there is only one instance of it,
			// or if all the instances have the same entering flag.
			j += (entering == 0 || entering == total);
		}
		intersections.resize(j);
	}
	
	// A "vertex" is a polygon vertex annotated with (possibly) a link to a
	// different vertex that is equivalent to it.
	class Vertex {
	public:
		Vertex() = default;
		Vertex(Point point, int link = -1) : point(point), link(link) {}
		
		Point point;
		int link = -1;
	};
}



// Translate a ring by the given vector.
Polygon Polygon::operator+(Point offset) const
{
	Polygon result = *this;
	result += offset;
	return result;
}



Polygon &Polygon::operator+=(Point offset)
{
	for(Ring &ring : *this)
		ring += offset;
	return *this;
}



Polygon Polygon::operator-(Point offset) const
{
	return *this + -offset;
}



Polygon &Polygon::operator-=(Point offset)
{
	return (*this += -offset);
}



// Scale a ring by the given amount.
Polygon Polygon::operator*(int scale) const
{
	Polygon result = *this;
	result *= scale;
	return result;
}



Polygon &Polygon::operator*=(int scale)
{
	for(Ring &ring : *this)
		ring *= scale;
	return *this;
}



Polygon operator*(int scale, const Polygon &poly)
{
	Polygon result = poly;
	result *= scale;
	return result;
}



Polygon Polygon::operator/(int scale) const
{
	Polygon result = *this;
	result /= scale;
	return result;
}



Polygon &Polygon::operator/=(int scale)
{
	for(Ring &ring : *this)
		ring /= scale;
	return *this;
}



// A "ring" is a simple polygon. If the points are in clockwise order it is
// a filled polygon; otherwise it is a hole.
void Polygon::Add(const Ring &ring)
{
	// In the comments below, this polygon is called P, and the simple polygon
	// being added is called Q. The output is called R.
	
	// We're going to generate a new output polygon here:
	Polygon result;
	
	// Remember all intersections with all edges on Q, so we can slot them in
	// at the right spots when generating the new Q.
	vector<Intersection> qIntersections;
	int qMaxOrder = MaxOrder(ring);
	
	// This is all the parts in one vector, with intersection vertices added
	// and links between polygons (including joining end to start points).
	vector<Vertex> amplified;
	
	// Remember the smallest part of P that contains Q.
	const Ring *smallestContainer = nullptr;
	
	// Entering and exiting calculations are reversed if Q is a hole.
	bool isHole = ring.IsHole();
	
	
	// Step 1: finding the intersections and generating copies of the P polygons
	// with those intersections inserted:
	
	// Loop through every segment of every part of this polygon, checking for
	// intersections with the given ring. If there are any, generate an
	// "amplified" version of the part with all the intersection vertices added.
	for(const Ring &part : *this)
	{
		// First, find any potential intersections. It's possible that we will
		// generate duplicates here, e.g. if an intersection occurs at a vertex.
		vector<Intersection> pIntersections;
		int pMaxOrder = MaxOrder(part);
		
		for(Edge p(part); p; ++p)
			for(Edge q(ring); q; ++q)
			{
				int cross = p.Vector().Cross(q.Vector());
				if(!cross)
					continue;
				
				// Now, we know the two lines are not parallel, so they must
				// intersect. Calculate where that intersection occurs.
				Point d = q.Start() - p.Start();
				int pT = d.Cross(q.Vector());
				int qT = d.Cross(p.Vector());
				// If the cross product is negative, negate all three so that
				// the comparisons below will work correctly.
				bool entering = (cross > 0) ^ isHole;
				if(cross < 0)
				{
					cross = -cross;
					pT = -pT;
					qT = -qT;
				}
				
				// Check if the intersection occurs within both segments,
				// including their endpoints. If so, record the intersection.
				if(qT >= 0 && qT <= cross && pT >= 0 && pT <= cross)
				{
					// The "order" for this intersection represents how far it
					// is around P. Order values must monotonically increase
					// along each edge and from one edge to the next, and the
					// order of the end of one edge should equal the order of
					// the beginning of the next.
					// 64-bit math is needed for this calculation.
					Point dP(
						static_cast<int64_t>(pT) * p.Vector().X() / cross,
						static_cast<int64_t>(pT) * p.Vector().Y() / cross);
					Point dQ(
						static_cast<int64_t>(qT) * q.Vector().X() / cross,
						static_cast<int64_t>(qT) * q.Vector().Y() / cross);
					pIntersections.emplace_back(
						p.Start() + dP,
						(p.Order() + dP.Dot(dP)) % pMaxOrder,
						(q.Order() + dQ.Dot(dQ)) % qMaxOrder,
						entering);
				}
			}
		// Sort the intersections so we can loop through them in order.
		Sort(pIntersections);
		// Remove any "doubles" in the list. (This also includes quadruples when
		// two polygons cross at their vertices.)
		RemoveDoubles(pIntersections);
		
		// If that removed all the intersections, this part does not intersect
		// the newly added ring.
		if(pIntersections.empty())
		{
			// Parts of P that do not intersect Q should be added to the output
			// as long as they are not entirely contained within Q.
			if(!ring.Contains(part))
			{
				result.emplace_back(part);
				
				// If Q doesn't contain this part of P, check if the opposite is
				// true. If this part contains Q, check if it's contained in any
				// other "container" we've found so far.
				
				// Anything that contains the ring must be either nested in this
				// part or around it; there cannot be any intersection because
				// the rings in P do not intersect.
				if(part.Contains(ring) && (!smallestContainer || smallestContainer->Contains(part)))
					smallestContainer = &part;
			}
			continue;
		}
		// Otherwise, we need to go back around the part to generate the
		// amplified list of vertices.
		
		// Remember the index in the "amplified" vector where this part begins.
		// We'll need that later to construct the closing loop.
		int pIndex = amplified.size();
		
		// If two intersections in a row have the same "entering" flag, ignore
		// the second one.
		bool wasEntering = pIntersections.back().entering;
		// Just to avoid always having to check if we're at the end, insert a
		// sentinel into the end of the candidates list that has a high enough
		// order that we'll never reach it.
		pIntersections.emplace_back(Point(), pMaxOrder);
		vector<Intersection>::const_iterator it = pIntersections.begin();
		for(Edge p(part); p; )
		{
			// Add the start point to the list of amplified vertices. But, don't
			// add it twice if there's an intersection at the start of this segment.
			if(it->order != p.Order())
				amplified.push_back(p.Start());
			// Advance p to the next segment, and then advance the intersection
			// iterator until it reaches the next segment, too.
			for(++p; it->order < p.Order(); ++it)
			{
				// If the edge of one polygon goes from outside to collinear
				// with the other, and then collinear to inside, that generates
				// two "entering" intersections, of which only one should count.
				if(it->entering == wasEntering)
					continue;
				wasEntering = it->entering;
				
				// The "index" field was used to store the Q order.
				qIntersections.emplace_back(it->point, it->index, amplified.size(), it->entering);
				amplified.emplace_back(it->point);
			}
		}
		// Close out the amplified polygon.
		amplified.emplace_back(amplified[pIndex].point, pIndex);
	}
	
	
	// Step 2: generating a copy of Q with the intersections inserted:
	
	// If there were no intersections at all:
	if(qIntersections.empty())
	{
		// Q should be added to P if it has the opposite polarity of the
		// smallest part of P that fully contains Q. (If none, the space outside
		// the polygon counts as having "hole" polarity.)
		bool containerIsHole = (!smallestContainer || smallestContainer->IsHole());
		if(containerIsHole != isHole)
			result.emplace_back(ring);
		
		result.swap(*this);
		return;
	}
	
	// Sort the intersections so we can loop through them in order. This time,
	// we don't need to check for "doubles" since they were removed above.
	Sort(qIntersections);
	
	// Remember the index in the "amplified" vector where this part begins.
	// We'll need that later to construct the closing loop.
	int qIndex = amplified.size();
	
	// Remember what indices in the amplified vector list are possible
	// starting points for a polygon trace.
	vector<int> enteringIndices;
	
	// Just to avoid always having to check if we're at the end, insert a
	// sentinel into the end of the candidates list that has a high enough
	// order that we'll never reach it.
	qIntersections.emplace_back(Point(), qMaxOrder);
	vector<Intersection>::const_iterator it = qIntersections.begin();
	// Loop through the edges of q.
	for(Edge q(ring); q; )
	{
		// Add the start point to the list of amplified vertices. But, don't
		// add it twice if there's an intersection at the start of this segment.
		if(it->order != q.Order())
			amplified.push_back(q.Start());
		// Advance q to the next segment, and then advance the intersection
		// iterator until it reaches the next segment, too.
		for(++q; it->order < q.Order(); ++it)
		{
			// If this is an "entering" intersection, store it as one of the
			// possible places to start a trace.
			if(it->entering)
				enteringIndices.emplace_back(amplified.size());
			// Add links in both directions.
			if(!it->entering)
				amplified[it->index].link = amplified.size();
			amplified.emplace_back(it->point);
			if(it->entering)
				amplified.back().link = it->index;
		}
	}
	// Close out the amplified polygon.
	amplified.emplace_back(amplified[qIndex].point, qIndex);
	
	
	// Step 3: Tracing the output polygons.
	
	// Try each possible starting index.
	for(int index : enteringIndices)
	{
		if(amplified[index].link < 0)
			continue;
		
		// Begin editing a new output polygon.
		result.emplace_back();
		Ring &out = result.back();
		
		int j = index;
		do {
			Vertex &vertex = amplified[j];
			
			// Follow the link, then disconnect it. This is to make sure we
			// only trace each polygon once.
			if(vertex.link >= 0)
			{
				j = vertex.link;
				vertex.link = -2;
			}
			else
			{
				out.emplace_back(vertex.point);
				++j;
			}
		} while(j != index);
		
		// Make sure we didn't create a degenerate polygon.
		if(out.size() < 3)
			result.pop_back();
	}
	result.swap(*this);
}



// Get just the component of this polygon that contains the given point.
// That includes any holes in that polygon.
void Polygon::FloodFill(Point point)
{
	vector<float> area(size(), 0.f);
	vector<bool> keep(size(), false);
	int keepCount = 0;
	for(size_t i = 0; i < size(); ++i)
		area[i] = (*this)[i].Area();
	
	// First, find the smallest positive part that contains the point.
	size_t smallestI = 0;
	for(size_t i = 0; i < size(); ++i)
		if(area[i] > 0.f && (*this)[i].Contains(point) && (!keepCount || area[i] < area[smallestI]))
		{
			keepCount = 1;
			smallestI = i;
		}
	// If the given point is not contained within a polygon, return an empty set.
	if(!keepCount)
	{
		clear();
		return;
	}
	keep[smallestI] = true;
	Ring &smallest = (*this)[smallestI];
	
	// Now, find all holes that are inside that polygon. But, make sure we do
	// not include holes that are nested inside other holes.
	for(size_t i = 0; i < size(); ++i)
		if(area[i] < 0.f && smallest.Contains((*this)[i]))
		{
			bool nested = false;
			// Find all holes larger than this one and see if they contain this
			// one. To contain it, the other hole must be larger, and since
			// holes have negative area that must mean a lower area.
			for(size_t j = 0; j < size(); ++j)
				if(area[j] < area[i] && (*this)[j].Contains((*this)[i].front()))
				{
					nested = true;
					break;
				}
			if(!nested)
				keep[i] = true;
		}
	Polygon result;
	result.reserve(keepCount);
	for(size_t i = 0; i < size(); ++i)
		if(keep[i])
		{
			result.emplace_back();
			result.back().swap((*this)[i]);
		}
	result.swap(*this);
}



// Check if this polygon contains the given point.
bool Polygon::Contains(Point point) const
{
	// Just add up the "winding number" of all the parts of this polygon and
	// check if it is nonzero. As a shortcut, if the point is on any edge of the
	// polygon, it must be inside.
	int winding = 0;
	for(const Ring &part : *this)
	{
		pair<int, int> result = part.Winding(point);
		if(result.second)
			return true;
		winding += result.first;
	}
	return winding;
}



// Check if the given line segment (excluding its endpoints) intersects this
// polygon. Touching a vertex counts only if the edge the vertex is on is
// not exactly collinear with the given line segment.
bool Polygon::Intersects(Point start, Point end) const
{
	Point qV = end - start;
	for(const Ring &part : *this)
		for(Edge p(part); p; ++p)
		{
			int cross = p.Vector().Cross(qV);
			if(!cross)
				continue;
			
			// Now, we know the two lines are not parallel, so they must
			// intersect. Calculate where that intersection occurs.
			Point d = start - p.Start();
			int pT = d.Cross(qV);
			int qT = d.Cross(p.Vector());
			// If the cross product is negative, negate all three so that
			// the comparisons below will work correctly.
			if(cross < 0)
			{
				cross = -cross;
				pT = -pT;
				qT = -qT;
			}
			
			// Check if the intersection occurs within both segments,
			// including their endpoints. If so, record the intersection.
			if(qT > 0 && qT < cross && pT >= 0 && pT <= cross)
				return true;
		}
	return false;
}

File: web/polygon.js

import {Point} from "./point.js";

/** A polygon, made up of one or more "rings" of vertices.
 * @typedef {Object} Polygon
 */
export class Polygon {
  /** Add a "ring" of points to this polygon, without calculating intersections.
   * @param {number[]} vertices
   */
  push(points) {
    if (points.length) this._rings.push(expand(points));
  }
  /** Check if this polygon is empty.
   * @returns {boolean} True if the polygon is empty.
   */
  isEmpty() {
    return !this._rings.length;
  }
  /** Add the given vector offset to this polygon.
   * @param {Point} offset
   * @returns {Polygon}
   */
  add(point) {
    const polygon = new Polygon();
    for (const ring of this._rings) {
      const result = [];
      for (let i = 0; i < ring.length; i += 4) {
        result.push(ring[i] + point.x, ring[i + 1] + point.y, ring[i + 2], ring[i + 3]);
      }
      polygon._rings.push(result);
    }
    return polygon;
  }
  /** Add the given Polygon to this polygon, one "ring" at a time.
   * @param {Polygon} polygon
   */
  union(polygon) {
    for (const ring of polygon._rings) {
      this._union(ring);
    }
  }
  _union(ring) {
    // In the comments below, this polygon is called P, and the simple polygon
    // being added is called Q. The output is called R.

    // We're going to generate a new output polygon here:
    const result = [];

    // Remember all intersections with all edges on Q, so we can slot them in
    // at the right spots when generating the new Q.
    const qIntersections = [];

    // This is all the parts in one vector, with intersection vertices added
    // and links between polygons (including joining end to start points).
    const amplified = [];

    // Remember the smallest part of P that contains Q.
    let smallestContainer;

    // Entering and exiting calculations are reversed if Q is a hole.
    const qIsHole = isHole(ring);

    // Step 1: finding the intersections and generating copies of the P polygons
    // with those intersections inserted.

    // Loop through every segment of every part of this polygon, checking for
    // intersections with the given ring. If there are any, generate an
    // "amplified" version of the part with all the intersection vertices added.
    for (const part of this._rings) {
	  	// First, find any potential intersections. It's possible that we will
	  	// generate duplicates here, e.g. if an intersection occurs at a vertex.
      const pIntersections = [];

      for (let pi = 0; pi < part.length; pi += 4) {
        for (let qi = 0; qi < ring.length; qi += 4) {
          const cross = part[pi + 2] * ring[qi + 3] - part[pi + 3] * ring[qi + 2];
          if (!cross) continue;

          // Now, we know the two lines are not parallel, so they must
          // intersect. Calculate where that intersection occurs.
          const dx = ring[qi] - part[pi];
          const dy = ring[qi + 1] - part[pi + 1];
          // pT and qT are between 0 and 1 if the intersection is on p and q
          // respectively. As long as the vertices are all integer coordinates,
          // the intersection point's coordinates will be divisible by 0.25, but
          // that does not hold true for pT and qT.
          const pC = dx * ring[qi + 3] - dy * ring[qi + 2];
          const pT = pC / cross;
          if (pT < 0 || pT > 1) continue;
          const qC = dx * part[pi + 3] - dy * part[pi + 2];
          const qT = qC / cross;
          if (qT < 0 || qT > 1) continue;
          
          // Avoid floating point rounding error by multiplying by pC first
          // before dividing by cross.
          const point = new Point(
            part[pi] + part[pi + 2] * pC / cross,
            part[pi + 1] + part[pi + 3] * pC / cross);
          pIntersections.push({
            point,
            order: (pi * 0.25 + pT) % (part.length * 0.25),
            qOrder: (qi * 0.25 + qT) % (ring.length * 0.25),
            entering: (cross > 0) != qIsHole
          });
        }
      }
      // Sort the intersections so we can loop through them in order.
      pIntersections.sort((a, b) => a.order - b.order);
      // Remove "doubles" due to intersections exactly at a vertex.
      removeDoubles(pIntersections);

      // If there are no intersections left, figure out if this part of the
      // polygon is entirely within the new ring.
      if (!pIntersections.length) {
        if (!ringContainsRing(ring, part)) {
          result.push(part);
          if (ringContainsRing(part, ring) && (!smallestContainer || ringContainsRing(smallestContainer, part))) {
            smallestContainer = part;
          }
        }
        continue;
      }

      // Otherwise, we need to go back around the part to generate the
      // amplified list of vertices.
      amplify(part, pIntersections, amplified, it => {
        qIntersections.push({
          point: it.point,
          order: it.qOrder,
          index: amplified.length,
          entering: it.entering
        });
        amplified.push(-1, it.point.x, it.point.y);
      });
    }

    // Step 2: generating a copy of Q with the intersections inserted:
    // If there were no intersections at all:
    if (!qIntersections.length) {
      // Q should be added to P if it has the opposite polarity of the
      // smallest part of P that fully contains Q. (If none, the space outside
      // the polygon counts as having "hole" polarity.)
      const containerIsHole = (!smallestContainer || isHole(smallestContainer));
      if (containerIsHole != qIsHole) result.push(ring);
      this._rings = result;
      return;
    }
      
    // Sort the intersections so we can loop through them in order. This time,
    // we don't need to check for "doubles" since they were removed above.
    qIntersections.sort((a, b) => a.order - b.order);
    // Iterate through the vertices, linking them up with previous parts.
    const enteringIndices = [];
    amplify(ring, qIntersections, amplified, it => {
      // If this is an "entering" intersection, store it as one of the
      // possible places to start a trace.
      if (it.entering) enteringIndices.push(amplified.length);
      // If this is an exiting vertex, add a link to it.
      else amplified[it.index] = amplified.length;
      // Otherwise, if this is an entering vertex, add a link from it.
      amplified.push(it.entering ? it.index : -1, it.point.x, it.point.y);
    });
  
  	// Step 3: Tracing the output polygons.
    // Try each possible starting index.
    for (const index of enteringIndices) {
      // Make sure this link has not been cleared (while tracing a previous output).
      if (amplified[index] < 0) continue;

      const out = tracePolygon(amplified, index);
      if (out.length >= 6) result.push(expand(out));
    }
    this._rings = result;
  }

  /** Remove every component of this polygon that is not reachable from the given point.
   * @param {Point} point
   */
  floodFill(point) {
    const areas = [];
    for (const ring of this._rings) {
      areas.push(area(ring));
    }

    // First, find the smallest positive ring that contains the point.
    let smallestI = -1;
    for (let i = 0; i < areas.length; i++) {
      if (areas[i] > 0 && contains(this._rings[i], point) && (smallestI < 0 || areas[i] < areas[smallestI])) {
        smallestI = i;
      }
    }
    // If we didn't find a positive ring containing the point, clear the whole polygon.
    if (smallestI < 0) {
      this._rings.length = 0;
      return;
    }

    // Begin building up the output polygon.
    const smallest = this._rings[smallestI];
    const result = [smallest];

    // Now, find all holes that are inside that polygon. But, make sure we do
    // not include holes that are nested inside other holes.
    outer:
    for (let i = 0; i < areas.length; ++i) {
      const ring = this._rings[i];
      if (areas[i] < 0 && ringContainsRing(smallest, ring)) {
        // Find all holes larger than this one and see if they contain this
		  	// one. To contain it, the other hole must be larger, and since
        // holes have negative area that must mean a lower area.
        for (let j = 0; j < areas.length; ++j) {
          if (areas[j] < areas[i] && ringContainsRing(this._rings[j], ring)) continue outer;
        }
        result.push(ring);
      }
    }
    this._rings = result;
  }

  /** Check if this polygon contains the given point.
   * @param {Point} point
   * @returns {boolean} True if the given point is in this polygon (or on an edge of it).
   */
  contains(point) {
    let winding = 0;
    for (const ring of this._rings) {
      const result = windingNumber(ring, point.x, point.y);
      if (result.border) return true;
      winding += result.winding;
    }
    return winding != 0;
  }

  /** Check if the given line segment intersects this polygon - that is, the segment has
   * portions both inside and outside the polygon. The polygon's border counts as inside,
   * so it is not an intersection if the segment touches the border but is otherwise
   * entirely inside the polygon.
   * @param {Point} start
   * @param {Point} end
   * @returns {boolean} True if the segment is not entirely inside or outside.
   */
  intersects(start, end) {
    const vector = end.sub(start);
    for (const ring of this._rings) {
      // The first edge will be between vertices 0 and 1. Calculate the vector of the
      // previous edge, i.e. from the back of the ring to 0.
      let prev = (ring[0] - ring[ring.length - 4]) * vector.y - (ring[1] - ring[ring.length - 3]) * vector.x;
      for (let i = 0; i < ring.length; i += 4) {
        const cross = ring[i + 2] * vector.y - ring[i + 3] * vector.x;
        const pCross = prev;
        prev = cross;
        if (start.x == ring[i] && start.y == ring[i + 1]) {
          if (pCross < 0 && cross < 0) return true;
          continue;
        }
        if (end.x == ring[i] && end.y == ring[i + 1]) {
          if (pCross > 0 && cross > 0) return true;
          continue;
        }
        // Ignore edges that are either parallel to or collinear with the query segment.
        // This is safe because if there is an intersection, it will be detected at one
        // of the adjacent edges.
        if (!cross) continue;
  
        // Now, we know the two lines are not parallel, so they must intersect. Calculate
        // where that intersection occurs.
        const dx = start.x - ring[i];
        const dy = start.y - ring[i + 1];
        const segT = (dx * ring[i + 3] - dy * ring[i + 2]) / cross;
        const edgeT = (dx * vector.y - dy * vector.x) / cross;
        // The T values are the solution to: intersection = start + T * vector. That is,
        // the intersection is on the segment if T is between 0 and 1.

        if (edgeT < 0 || edgeT > 1 || segT < 0 || segT > 1) continue;
        // If the intersection point is strictly inside the query segment (not at one of
        // its endpoints), it always counts as an intersection.
        if (segT > 0 && segT < 1) return true;
        // If we're here, segT is exactly 0 or 1, i.e. there's an intersection at an
        // endpoint. If we're not at an endpoint of the polygon, the segment is outside
        // the polygon if segT is 0 and cross < 0, or segT is 1 and cross > 0.
        if (edgeT !== 0 && edgeT !== 1 && !segT === (cross < 0)) return true;
      }
    }
    return false;
  }

  /** Access the array of Ring components that make up this polygon.
   * @returns {number[][]} An array of rings.
   */
  rings() {
    return this._rings;
  }

  /** @type {number[][]} */
  _rings = [];
}


function removeDoubles(intersections) {
  let j = 0;
  for (let i = 0; i < intersections.length; ) {
    intersections[j] = intersections[i];
    const order = intersections[i].order;
    let entering = intersections[i].entering;
    let total = 1;
    for (i++; i < intersections.length && intersections[i].order == order; i++) {
      entering += intersections[i].entering;
      total++;
    }
    // Keep intersections[j] only if every intersection point at that
    // same location has the same "entering" value. This is to exclude
    // places where a corner of one polygon touches but does not cross
    // an edge of the other.
    if (entering == 0 || entering == total) j++;
  }
  // Erase any intersections that were removed.
  intersections.length = j;
  if (!intersections.length) return;

  // Now, also trim any adjacent points with the same "entering" value.
  j = 0;
  let wasEntering = intersections[intersections.length - 1].entering;
  for (let i = 0; i < intersections.length; i++) {
    intersections[j] = intersections[i];
    if (intersections[i].entering != wasEntering) {
      wasEntering = !wasEntering;
      j++;
    }
  }
  intersections.length = j;
}


function amplify(vertices, intersections, amplified, fun) {
  // Remember the index in the "amplified" vector where this part begins.
  // We'll need that later to construct the closing loop.
  const start = amplified.length;
  // When the pOrder reaches this value, we're at a new vertex.
  let index = 0;
  for (const it of intersections) {
    // Output any vertices that are before this intersection.
    for ( ; index * 0.25 <= it.order; index += 4) {
      if (index * 0.25 !== it.order) amplified.push(-1, vertices[index], vertices[index + 1]);
    }
    // Process this intersection.
    fun(it);
  }
  // Add the remaining vertices to the amplified list.
  for ( ; index < vertices.length; index += 4) {
    amplified.push(-1, vertices[index], vertices[index + 1]);
  }
  // Close out the amplified polygon by adding a "link" from the
  // last vertex back to the first one.
  amplified.push(start, 0, 0);
}


function tracePolygon(amplified, index) {
  let j = index;
  const out = [];
  do {
    const link = amplified[j];
    if (link >= 0) {
      amplified[j] = -1;
      j = link;
    }
    else {
      out.push(amplified[j + 1], amplified[j + 2]);
      j += 3;
    }
  } while (j != index);
  return out;
}


// Turn a set of vertices into a set of vertices plus edge vectors.
function expand(points) {
  const ring = [];
  let startX = points[0];
  let startY = points[1];
  for (let i = 2; i < points.length; i += 2) {
    const endX = points[i];
    const endY = points[i + 1];
    ring.push(startX, startY, endX - startX, endY - startY);
    startX = endX;
    startY = endY;
  }
  ring.push(startX, startY, points[0] - startX, points[1] - startY);
  return ring;
}


// Check if a ring contains a point.
function contains(ring, point) {
  const result = windingNumber(ring, point.x, point.y);
  return (result.winding || result.border);
}


// Get the area of a ring.
function area(ring) {
  // Break the polygon up into triangles and add up their areas. The triangles
  // are each formed using an edge of this ring as one edge, and connecting
  // them all to the same arbitrary point as the third vertex. That vertex
  // does not even need to be inside the polygon, so simplify the math by just
  // using the origin as the third point.
  
  // The area of a triangle is half the cross product of the two vectors
  // coming out of any of the vertices.
  let area = 0;
  for (let i = 0; i < ring.length; i += 4) {
    // The cross product is: startX * endY - startY * endX
    //   = x * (y + dy) - y * (x + dx)
    //   = x * y + x * dy - x * y - dx * y
    //   = x * dy - dx * y
    area += ring[i] * ring[i + 3] - ring[i + 1] * ring[i + 2];
  }
  return 0.5 * area;
}

function isHole(ring) {
  return area(ring) < 0;
}

// This helper function is only for use by Polygon. Given a ring that does not
// intersect this ring, check if it is inside this one.
function ringContainsRing(ring, other) {
  const result = windingNumber(ring, other[0], other[1]);
  return result.winding && (!result.border || area(ring) > area(other));
}

// Get the "winding number" and number of boundary intersections for the given point.
// This is returned as a {winding, border} object.
function windingNumber(ring, px, py) {
  let winding = 0;
  let border = 0;
  for (let i = 0; i < ring.length; i += 4) {
    // Special case: if this is a horizontal edge, the "border" value still needs to
    // be updated.
    const x = ring[i];
    const y = ring[i + 1];
    const dx = ring[i + 2];
    const dy = ring[i + 3];
    if (!dy) {
      if (y === py) border += (x + Math.min(0, dx) <= px && px <= x + Math.max(0, dx));
      continue;
    }
    const startsBelow = y <= py;
    const endsBelow = y + dy <= py;
    if (startsBelow == endsBelow) continue;
    
    const cross = (px - x) * dy - (py - y) * dx;
    winding += (!endsBelow && cross <= 0);
    winding -= (endsBelow && cross > 0);
    border += !cross;
  }
  return {winding, border};
}
